Add 3-letter language code support

jcsg opened the pull request:
This PR adds basic support for three-letter language and country codes. Using the existing data in `languages.json` and `countries.json`, data in the language fields `639_2B` and `639_2T` along with data in the country field `iso3` is added into the `languageref.json` and `countryref.json` files. This allows three new methods to function:

1. `.getTwoLetterLanguage()` - Given a three-letter language code, this returns its two-letter equivalent, if it exists, otherwise it returns `false`.

2. `.getTwoLetterCountry()` - Given a three-letter country code, this returns its two-letter equivalent. In the data, both the two-letter and three-letter codes are always given, so this does not guard against the possibility of either missing.

3. `.getThreeLetterCountry()` - Given a two-letter country code, this returns its three-letter equivalent. In the data, both the two-letter and three-letter codes are always given, so this does not guard against the possibility of either missing.
 
4. `.getThreeLetterLanguage()` - There are **2** possible three-letter language codes for 20 countries. ISO 639_2B specifies a 'bibliographic' code that is English-based, while ISO 639_2T specifies a  'terminological' code derived from the native name for the language. This latter T-code is generally preferred, including in future ISO standards. As an example, the language codes for German are:  
639_2B == 'ger' (from English *German*)  
639_2T == 'deu' (from German *Deutsch*)  
By default, this function returns the T-code if available. However, this can be overridden by adding `true` after the language code to fill the optional boolean parameter `prefer_b`.

After implementing these methods and modifying the `countryref.json` and `languageref.json` fields, the `.getAllLanguagesLike()` method was returning the right results, but not with an exact match first. I modified this method to make sure that exact matches were returned first in the list.

After running `npm test` on my branch, all 140 tests pass. 

I don't want this to break anything in Carmen ([like possibly here](https://github.com/mapbox/carmen/blob/master/lib/util/closest-lang.js#L6-L23)), but would like to be able to handle 3-letter language codes to enable better fallback behavior for directions.

cc: @mapbox/geocoding-gang

jcsg committed:
Added attempt at support for three-letter language and country codes.

jcsg committed:
Slightly modified getTwoLetterLanguage method to handle cases where no two-letter code exists.

jcsg committed:
Tried to fix problem that is causing Travis build to fail.

jcsg committed:
Tried to fix problem that is causing Travis build to fail.

jcsg committed:
Modifying getThreeLetterLanguage to try to pass Travis build check.

jcsg committed:
Line 132: changed let to var to try to get Travis to build

apendleton commented:
Carmen is pinned to a specific model-un version, so I wouldn't worry too much about that. If there are changes, we can always update later. To give more context around carmen's specific use: we essentially reimplement a couple of the core functions in model-un to have non-spec-conformant behavior that better fits our usecase; specifically, we use position rather case to infer the function of tags. In `model-un`, `getLanguage('fr')` returns a language whereas `getLanguage('FR')` returns a region; in carmen we always assume the subtag in the first position is a language and that all subsequent tags are not, so both would return the language.

As for your specific changes: I'm excited about the opportunity to support languages not tagged in 639-1 in the geocoder, so I fully support adding three-letter codes. I'm sort of surprised, though, by the sourcing of three-letter codes from ISO 639-2. My understanding is that it's no longer maintained, weirdly anglocentric, and has largely been replaced by ISO 639-3 (and, to some extent, 639-5), both of which have strong governance and better coverage. What's the rationale here? I know @boblannon has opinions about which specs to use here, and I think maybe @1ec5 does as well?

For code changes:
* `getLanguage` does not return a language code, it returns a language structure, complete with metadata. Given that, adding `getXLanguage` methods that return language codes rather than language structures is surprising from an ergonomics perspective, as there's precedent for `Language` to refer to an object. I'd have expected either for the functions to return objects, or for them to be called, say, `getThreeLetterLanguageCode`.
* It'd also be good to add some tests of `getLanguage` in addition to `hasLanguage` in conjunction with the three-letter codes being added.
* Lastly: having these new functions be case-insensitive is inconsistent with patterns used elsewhere in the library, where case sensitivity is assumed. I wouldn't actually necessarily be opposed to making the whole library case-insensitive (it would certainly make carmen's usecase easier as per above), but we shouldn't do it piecemeal, so absent a broader push in that direction I think ensuring lowercase should be the responsibility of the caller.

These are pretty minor tweaks though; no objection to the overall direction.

apendleton commented:
I should also say: for the geocoding usecase, at least, and I assume for directions as well, the vast majority of language codes we're going to receive are going to be derived by applications from the end user's locale, and there may be minor differences from platform to platform, but our focus should be on supporting languages that come from one of a couple of known, pretty small sets (Android locales, Apple locales, maybe something additional for browsers). Few if any developers can be expected to make up language codes to submit. Given that, I don't think we need to bend over backwards to support unexpected or esoteric codes or formattings. If no major platform uses 639-2B codes, for example, I don't see a strong motivation to support them (and I'm reasonably sure that's true, but would welcome counterexamples).

I'm interested in this change because it lets us support languages we just can't support at all right now because they have no 639-1 code, and because platforms in the wild (iOS, I think?) do actually use these three-letter codes instead of two-letter ones for a specific subset of languages that only have the former and not the latter. I don't personally see a strong need for expanding the scope of a change like this much beyond that, though. Even for cases like `eng`, I suppose we may as well support it if we're supporting a broader spec that includes it anyway, but I doubt we'll ever see it submitted.

jcsg commented:
Thanks for your comments, @apendleton! I'm glad it shouldn't interfere with Carmen.

I actually only sourced three-letter codes from 639-2 because they were already in the `languages.json` file. I can add the ones from 639-3. For what it's worth, the 639-2T codes, which are less anglocentric, are the ones that are preferred in the `getThreeLetterLanguage` function (which I should rename, to conform with the convention you mentioned about `Language` vs. `LanguageCode`).

Since I'd like to use `model-un` for processing language-locale codes that the directions API receives, I'll take a look at what it can do with the language and locale codes supported by Android, iOS, and at least Chrome, Safari, and Firefox. I'll modify this branch and ping everyone when it's in a better state to be merged.

apendleton commented:
My mistake for not catching that you were using codes that were already there! I still do think longer-term, -3 is probably a better bet, but sorry for scope-creeping on you.

1ec5 commented:
> platforms in the wild (iOS, I think?) do actually use these three-letter codes instead of two-letter ones for a specific subset of languages that only have the former and not the latter

Apple platforms use a three-letter T code as a fallback when a two-letter code is unavailable for the user’s preferred language. Currently, iOS and macOS are only localized into languages that have two-letter codes, but the user can still choose a preferred language from a much larger list (maybe based on CLDR?) that includes everything from Filipino (`fil`) to Ainu (`ain`) to Klingon (`tlh`). Recent versions of iOS and macOS follow ISO 639-3 as far as treating certain ISO 639-2 codes as macrolanguages. For example, the language options omit Kurdish (`ku`) in favor of Sorani (`ckb`).

In Safari and possibly other [browsers](https://www.w3.org/International/questions/qa-lang-priorities), the `navigator.language` property used by Web applications reflects the system preference exactly the way a native iOS or macOS application does. Firefox lets the user set a custom language tag, but the interface is well hidden so we don’t have to worry about it.

boblannon commented:
sorrym, just now catching up on this.  looks great!  echoing @apendleton:

> I'm interested in this change because it lets us support languages we just can't support at all right now because they have no 639-1 code

jcsg committed:
Function names now reflect internal division of language vs. language code. Additional tests.

jcsg committed:
Refined test.js.

jcsg committed:
Added hasCountry test for iso-3 code.

jcsg commented:
**Updates**  
* Changed function names from `...Language` to `LanguageCode` to conform to established conventions already present in the original code.
* Removed case-insensitivity in language code processing.  
*NB*: It appears that country codes actually *should* be lowercased to be found in the `data/countryref.json` file. This processing is already being done in the `update.js` file currently in the master branch.
* Added tests as @apendleton suggested for `getLanguage`, plus more for `getCountry` and `hasCountry`. Also added initial basic tests for `getTwoLetterCountryCode` and `getThreeLetterCountryCode`.

cc: @mapbox/geocoding-gang @boblannon @1ec5
